test/controllers/contacts_controller_test.rb

```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
+
+require 'test_helper'
+
+class ContactsControllerTest < ActionDispatch::IntegrationTest
+  test 'should get new' do
+    get contacts_path
+    assert_response :success
+  end
+
+  test 'Create Contact' do
+    post contacts_url,
+         params: { contact: { name: 'Johannes kann es', email: 'KannEs@Johannes.de',
+                              message: 'i am jo---han---es and show me what your got' } }
+    assert Contact.find_by_email('KannEs@Johannes.de')
+  end
+
+  test 'Check validation from message length Contact' do
+    post contacts_url,
+         params: { contact: { name: 'Johannes kann es', email: 'KannGarnichtsEs@Johannes.de', message: 'i ' } }
+    assert_not Contact.find_by_email('KannGarnichtsEs@Johannes.de')
+  end
+
+  test 'Check validation from email Contact' do
+    post contacts_url,
+         params: { contact: { name: 'Johannes kann es', email: 'KannGarnichtsEsJohannes.de',
+                              message: 'Ich möchte mich Bewwerben TROLLLLLLLL' } }
+    assert_not Contact.find_by_email('KannGarnichtsEsJohannes.de')
+  end
+
+  test 'Check validation from name Contact' do
+    post contacts_url,
+         params: { contact: { name: 'Jo', email: 'Peter@Johannes.de',
+                              message: 'Ich möchte mich Bewwerben TROLLLLLLLL' } }
+    assert_not Contact.find_by_email('Peter@Johannes.de')
+  end
```

Comment:
Das ist ein ganz typischer Unit-Test. Du schreibst es aber hier in einem `ActionDispatch::IntegrationTest`. Dafür wird die gesamte Anwendung simuliert, was sehr kostspielig ist. Dabei testest du hier keine Integration. Wenn du noch vielleicht UI-Nachrichten oder Button-Zustände abprüfen würdest, wäre das hier so angebracht. Sonst aber verschiebe das in einen Unit-Test.

Reply:
Alles klar dann werde ich das anpassen

Reply:
> Das ist ein ganz typischer Unit-Test. Du schreibst es aber hier in einem `ActionDispatch::IntegrationTest`. Dafür wird die gesamte Anwendung simuliert, was sehr kostspielig ist. Dabei testest du hier keine Integration. Wenn du noch vielleicht UI-Nachrichten oder Button-Zustände abprüfen würdest, wäre das hier so angebracht. Sonst aber verschiebe das in einen Unit-Test.

Die Tests wurden doch teils so generiert, warum sollen diese jetzt extra dafür umgeschrieben werden?
außerdem, was ist wenn der Controller im vollem von Sachen wie dem ApplicationController und deren Methoden abhängen? Man kann ja nicht alles in der Umgebung mocken...

Reply:
Das was generiert wurde, ist ein Controller-Test und ist korrekt. Da wird aber noch nichts getestet.

Die Zeilen 21, 28 und 35 sind dann aber Tests, die die Validierung testen. Das gehört in ein Model-Test. Dafür gibt es eigene "Gruppe" und auch Unterordner. Dafür braucht man dann auch nichts mocken.

Wie immer, empfehle ich euch auch hierzu die Rails-Guides:
[Functional Tests for Your Controllers](https://guides.rubyonrails.org/testing.html#functional-tests-for-your-controllers)
[Model Testing](https://guides.rubyonrails.org/testing.html#model-testing)

Reply:
> Das was generiert wurde, ist ein Controller-Test und ist korrekt. Da wird aber noch nichts getestet.
> 
> Wie immer, empfehle ich euch auch hierzu die Rails-Guides:
> [Functional Tests for Your Controllers](https://guides.rubyonrails.org/testing.html#functional-tests-for-your-controllers)
> 
>     * was the web request successful?
> 
>     * was the user redirected to the right page?
> 
>     * was the user successfully authenticated?
> 
>     * was the appropriate message displayed to the user in the view?
> 
>     * was the correct information displayed in the response?
> 
> 
> Die Zeilen 21, 28 und 35 sind dann aber Tests, die die Validierung testen. Das gehört in ein Model-Test. Dafür gibt es eigene "Gruppe" und auch Unterordner. Dafür braucht man dann auch nichts mocken.
> 
> [Model Testing](https://guides.rubyonrails.org/testing.html#model-testing)
> 
>     * Model tests are used to test the various models of your application.
> 
> 
> Einfach gesagt, testet man im Model-Tests das, was im Model gemacht wird. Im Controller-Test ebenso. Im Model schreibt ihr Validierung. Daher wird diese nicht im Controller-Test getestet.

aaaaaaaah, vielen dank ~